src/Sentry/MetricAggregator.cs

```diff
@@ -0,0 +1,459 @@
+    /// An enumerable containing the keys for any buckets that are ready to be flushed
+    /// </returns>
+    internal IEnumerable<long> GetFlushableBuckets(bool force = false)
+    {
+        if (!_buckets.IsValueCreated)
+        {
+            yield break;
+        }
+
+        if (force)
+        {
+            // Return all the buckets in this case
+            foreach (var key in Buckets.Keys)
+            {
+                yield return key;
+            }
+        }
+        else
+        {
+            var cutoff = MetricHelper.GetCutoff();
+            foreach (var key in Buckets.Keys)
+            {
+                var bucketTime = DateTimeOffset.FromUnixTimeSeconds(key);
+                if (bucketTime < cutoff)
+                {
+                    yield return key;
+                }
+            }
+        }
+    }
+
+    private Dictionary<long, Dictionary<MetricResourceIdentifier, SentryStackFrame>> FlushableLocations()
+    {
+        _codeLocationLock.Wait();
+        try
+        {
+            var result = _pendingLocations;
+            _pendingLocations = new Dictionary<long, Dictionary<MetricResourceIdentifier, SentryStackFrame>>();
+            return result;
+        }
+        finally
+        {
+            _codeLocationLock.Release();
+        }
+    }
+
+    /// <summary>
+    /// Clear out stale seen locations once a day
+    /// </summary>
+    private void ClearStaleLocations()
+    {
+        var now = DateTimeOffset.UtcNow;
+        var today = now.GetDayBucketKey();
+        if (lastClearedStaleLocations == today)
+        {
+            return;
+        }
+        // Allow 60 seconds for all code locations to be sent at the transition from one day to the next
+        const int staleGraceInMinutes = 1;
+        if (now.Minute < staleGraceInMinutes)
+        {
+            return;
+        }
+
+        foreach (var dailyValues in _seenLocations.Keys.ToArray())
+        {
+            if (dailyValues < today)
+            {
+                _seenLocations.TryRemove(dailyValues, out _);
+            }
+        }
+        lastClearedStaleLocations = today;
+    }
+
+    /// <inheritdoc cref="IAsyncDisposable.DisposeAsync"/>
+    public async ValueTask DisposeAsync()
+    {
+        _options.LogDebug("Disposing MetricAggregator.");
+
+        if (_disposed)
+        {
+            _options.LogDebug("Already disposed MetricAggregator.");
+            return;
+        }
+
+        _disposed = true;
+
+        try
+        {
+            // Request the LoopTask stop.
+            await _shutdownSource.CancelAsync().ConfigureAwait(false);
+
+            // Now wait for the Loop to stop.
+            // NOTE: While non-intuitive, do not pass a timeout or cancellation token here.  We are waiting for
+            // the _continuation_ of the method, not its _execution_.  If we stop waiting prematurely, this may cause
+            // unexpected behavior in client applications.
+            LoopTask.Wait();
```

Comment:
this is a blocking call, did you mean:
```suggestion
            await LoopTask;
```
?